@ehpor unfortunately I get stuck at this:
```
(test) maintenncesMBP7:~ ilaginja$ mtop catkit_header
/Users/ilaginja/opt/miniconda3/envs/test/lib/python3.9/site-packages/asdf/asdf.py:10: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.
  from pkg_resources import parse_version
Traceback (most recent call last):
  File "/Users/ilaginja/opt/miniconda3/envs/test/bin/mtop", line 8, in <module>
    sys.exit(main())
  File "/Users/ilaginja/repos/catkit2/catkit2/tui/mtop.py", line 158, in main
    header = SharedMemory.open(shared_memory_id)
RuntimeError: Something went wrong while opening shared memory: No such file or directory
(test) maintenncesMBP7:~ ilaginja$ 
```

--- a/catkit2/testbed/testbed.py
+++ b/catkit2/testbed/testbed.py
@@ -266,6 +266,12 @@ class Testbed:
         if self.is_simulated:
             self.startup_services.append('simulator')
 
+        # Create a message broker.
+        self.message_broker_header = create_shared_memory(f'catkit_broker_{port}.hdr', 1024 * 1024 * 1024)
+        self.message_broker_buffer = create_shared_memory(f'catkit_broker_{port}.buf', 1024 * 1024 * 1024 * 2)
+
+        self.message_broker = LocalMessageBroker.create(self.message_broker_header, [self.message_broker_buffer])
+
         # Fill in services dictionary.
         for service_id, service_info in self.config['services'].items():
             service_type = service_info['service_type']
@@ -345,12 +351,6 @@ class Testbed:
         self.shutdown_flag = threading.Event()
 
         self.heartbeat_stream = DataStream.create('heartbeat', 'testbed', 'uint64', [1], 20)
-
-        # Create a message broker.
-        self.message_broker_header = create_shared_memory(f'catkit_broker_{port}.hdr', 1024 * 1024 * 1024)
-        self.message_broker_buffer = create_shared_memory(f'catkit_broker_{port}.buf', 1024 * 1024 * 1024 * 2)
-
-        self.message_broker = LocalMessageBroker.create(self.message_broker_header, [self.message_broker_buffer])
 
     def run(self):
         '''Run the main loop of the server.

--- a/catkit2/testbed/testbed.py
+++ b/catkit2/testbed/testbed.py
@@ -111,9 +111,10 @@ class ServiceReference:
     state : ServiceState
         The current state of the service.
     '''
-    def __init__(self, service_id, service_type, state, dependencies):
+    def __init__(self, service_id, service_type, state, dependencies, broker):
         self.service_id = service_id
         self.service_type = service_type
+        self.broker = broker
 
         if dependencies is None:
             dependencies = []
@@ -140,6 +141,8 @@ class ServiceReference:
     def state(self, state):
         new_state = np.array([state.value], dtype='int8')
         self.state_stream.submit_data(new_state)
+
+        self.broker.publish_array(f'{self.service_id}/service_state/get', new_state)
 
     @property
     def is_alive(self):
@@ -272,7 +275,7 @@ class Testbed:
 
             dependencies = service_info.get('depends_on', [])
 
-            self.services[service_id] = ServiceReference(service_id, service_type, ServiceState.CLOSED, dependencies)
+            self.services[service_id] = ServiceReference(service_id, service_type, ServiceState.CLOSED, dependencies, self.message_broker)
 
         # Set up dependency management.
         for service_id, service in self.services.items():
@@ -429,6 +432,8 @@ class Testbed:
 
             heartbeat = np.array([get_timestamp()], dtype='uint64')
             self.heartbeat_stream.submit_data(heartbeat)
+
+            self.message_broker.publish_array('testbed/heartbeat/get', heartbeat)
 
     def monitor_services(self):
         while not self.shutdown_flag.is_set():

--- a/catkit2/tui/mtop.py
+++ b/catkit2/tui/mtop.py
@@ -84,7 +84,7 @@ class MTopApp(App):
         self.current_sort_column = 'topic'
 
     def compose(self) -> ComposeResult:
-        yield Header(icon='')
+        yield Header()
 
         self.table = DataTable(id="datatable")
         yield Container(self.table)

--- a/catkit_core/Service.cpp
+++ b/catkit_core/Service.cpp
@@ -131,6 +131,9 @@ void Service::Run(void (*error_check)())
 		std::uint64_t timestamp = GetTimeStamp();
 		m_Heartbeat->SubmitData(&timestamp);
 
+		ArrayInfo info{'u', '=', 8, 1, {1, 1, 1, 1}, {8, 1, 1, 1}};
+		m_Testbed->GetMessageBroker()->PublishArray(m_ServiceId + "/heartbeat/get", {info, &timestamp});
+
 		// Start the safety and heartbeat threads.
 		std::thread safety(&Service::MonitorSafety, this);
 		std::thread heartbeat(&Service::MonitorHeartbeats, this);
@@ -210,6 +213,10 @@ void Service::Run(void (*error_check)())
 	// Set heartbeat timestamp to zero to signal a dead service.
 	std::uint64_t timestamp = 0;
 	m_Heartbeat->SubmitData(&timestamp);
+
+	ArrayInfo info{'u', '=', 8, 1, {1, 1, 1, 1}, {8, 1, 1, 1}};
+	m_Testbed->GetMessageBroker()->PublishArray(m_ServiceId + "/heartbeat/get", {info, &timestamp});
+
 
 	CleanupAttributes();
 }
@@ -296,6 +303,9 @@ void Service::MonitorHeartbeats()
 		// Update my own heartbeat.
 		std::uint64_t timestamp = GetTimeStamp();
 		m_Heartbeat->SubmitData(&timestamp);
+
+		ArrayInfo info{'u', '=', 8, 1, {1, 1, 1, 1}, {8, 1, 1, 1}};
+		m_Testbed->GetMessageBroker()->PublishArray(m_ServiceId + "/heartbeat/get", {info, &timestamp});
 
 		// Check the testbed heartbeat.
 		if (!m_Testbed->IsAlive())

